File: src/controllers/task/TaskController.ts
import { TaskDTO, ITaskService } from './Task';

/**
 * Controller for Task CRUD operations
 */
export class TaskController {
  private taskService: ITaskService;

  constructor(taskService: ITaskService) {
    this.taskService = taskService;
  }

  /**
   * Create a new task
   */
  public async create(dto: TaskDTO): Promise<TaskCreateUpdateResponse> {
    const created = await this.taskService.create(dto);
    return {
      id: created.id,
      workspaceId: created.workspaceId,
      title: created.title,
      status: created.status,
      priority: created.priority,
      due: created.due,
    };
  }

  /**
   * Get a task by ID
   */
  public async getById(id: number): Promise<TaskCreateUpdateResponse | null> {
    const task = await this.taskService.getById(id);
    if (!task) return null;
    return {
      id: task.id,
      workspaceId: task.workspaceId,
      title: task.title,
      status: task.status,
      priority: task.priority,
      due: task.due,
    };
  }

  /**
   * Update an existing task
   */
  public async update(id: number, dto: TaskDTO): Promise<TaskCreateUpdateResponse | null> {
    const updated = await this.taskService.update(id, dto);
    if (!updated) return null;
    return {
      id: updated.id,
      workspaceId: updated.workspaceId,
      title: updated.title,
      status: updated.status,
      priority: updated.priority,
      due: updated.due,
    };
  }

  /**
   * Delete a task by ID
   */
  public async delete(id: number): Promise<boolean> {
    return await this.taskService.delete(id);
  }

  /**
   * Get all tasks
   */
  public async getAll(): Promise<TaskCreateUpdateResponse[]> {
    const tasks = await this.taskService.getAll();
    return tasks.map(task => ({
      id: task.id,
      workspaceId: task.workspaceId,
      title: task.title,
      status: task.status,
      priority: task.priority,
      due: task.due,
    }));
  }

  /**
   * Get tasks by workspace ID
   */
  public async getByWorkspaceId(workspaceId: number): Promise<TaskCreateUpdateResponse[]> {
    const tasks = await this.taskService.getByWorkspaceId(workspaceId);
    return tasks.map(task => ({
      id: task.id,
      workspaceId: task.workspaceId,
      title: task.title,
      status: task.status,
      priority: task.priority,
      due: task.due,
    }));
  }
}

File: src/main/java/com/example/demo/controller/task/TaskController.java
package com.example.demo.controller.task;

import java.util.List;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import com.example.demo.dto.task.TaskCreateRequest;
import com.example.demo.entity.task.Task;
import com.example.demo.service.task.TaskService;

import jakarta.validation.Valid;

@RestController
@RequestMapping("/api/tasks")
public class TaskController {

    private final TaskService taskService;

    public TaskController(TaskService taskService) {
        this.taskService = taskService;
    }

    @GetMapping
    public List<Task> getAll() {
        return taskService.getAll();
    }

    @GetMapping("/{id}")
    public ResponseEntity<Task> getById(@PathVariable Long id) {
        return taskService.getById(id)
                .map(ResponseEntity::ok)
                .orElseGet(() -> ResponseEntity.notFound().build());
    }

    @PostMapping
    public ResponseEntity<Task> create(@Valid @RequestBody TaskCreateRequest request) {
        Task created = taskService.create(request);
        return ResponseEntity.status(HttpStatus.CREATED).body(created);
    }

    @PutMapping("/{id}")
    public ResponseEntity<Task> update(@PathVariable Long id, @Valid @RequestBody TaskCreateRequest request) {
        return taskService.update(id, request)
                .map(ResponseEntity::ok)
                .orElseGet(() -> ResponseEntity.notFound().build());
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<Void> delete(@PathVariable Long id) {
        if (!taskService.delete(id)) {
            return ResponseEntity.notFound().build();
        }
        return ResponseEntity.noContent().build();
    }
}
